dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -789,15 +813,63 @@ BlockInputStreams MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreams(
     return res;
 }
 
+BlockInputStreams MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreamsPKOrder(
```

Comment:
I tried to hardcode `read_in_pk_order=true` at fetching columns to force execute this part of code and it seems that order of reading using this pipeline is still nondeterministic.